[7.1.0] Update Kafka-Connector doc with Avro Message support (#2783)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/kafka-connector/3.1.x/kafka-connector-avro-producer-example.md

# Avro Message with Kafka Connector Example

Given below is a sample scenario that demonstrates how to send Apache Avro messages to a Kafka broker via Kafka topics. The publishMessages operation allows you to publish messages to the Kafka brokers via Kafka topics.

## What you'll build

> Given below is a sample API that illustrates how you can connect to a Kakfa broker with the `init` operation and then use the `publishMessages` operation to publish messages via the topic. It exposes Kakfa functionalities as a RESTful service. Users can invoke the API using HTTP/HTTPs with the required information.

Reviewer: ```suggestion
Given below is a sample API that illustrates how you can connect to a Kafkaa broker with the `init` operation and then use the `publishMessages` operation to publish messages via the topic. It exposes Kakfa functionalities as a RESTful service. Users can invoke the API using HTTP/HTTPs with the required information.
```

Author: Should not this be Kafka?

> API has the context `/publishMessages`. It will publish messages via the topic to the Kafka server.

Reviewer: ```suggestion
API has the `/publishMessages` context. It publishes messages via the topic to the Kafka server.
```

            </inSequence>
            <outSequence/>
            <faultSequence/>
        </resource>
    </api>
    ```
Now we can export the imported connector and the API into a single CAR application. The CAR application needs to be deployed during server runtime. 

{!references/connectors/exporting-artifacts.md!}

## Deployment

Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 

{!references/connectors/deploy-capp.md!}
    
## Testing

Invoke the API (http://localhost:8290/publishMessages) with the following payload,

````json
{
   "test": {
       "type": "record",
       "name": "myrecord",
       "fields": [
           {
               "name": "f1",
               "type": ["string", "int"]
           }
       ]
   },
   "value": {
       "f1": "sampleValue"
   },
   "key": "sampleKey",
   "topic": "myTopic"
}
````

**Expected Response**: 
   
Run the following command to verify the messages:
````bash
[confluent_home]/bin/kafka-avro-console-consumer.sh --topic myTopic --bootstrap-server localhost:9092 --property print.key=true --from-beginning
````
See the following message content:
````json
{"f1":{"string":"sampleValue"}}
````  

> Sample API configuration when Confluent Schema Registry is secured with basic auth,

Reviewer: ```suggestion
Sample API configuration when the Confluent Schema Registry is secured with basic auth,
```

                      expression="json-eval($.topic)"
                      scope="default"
                      type="STRING"/>
            <kafkaTransport.init>
               <name>Sample_Kafka</name>
               <bootstrapServers>localhost:9092</bootstrapServers>
               <keySerializerClass>io.confluent.kafka.serializers.KafkaAvroSerializer</keySerializerClass>
               <valueSerializerClass>io.confluent.kafka.serializers.KafkaAvroSerializer</valueSerializerClass>
               <schemaRegistryUrl>http://localhost:8081</schemaRegistryUrl>
               <maxPoolSize>100</maxPoolSize>
               <basicAuthCredentialsSource>USER_INFO</basicAuthCredentialsSource>
               <basicAuthUserInfo>admin:admin</basicAuthUserInfo>
            </kafkaTransport.init>
            <kafkaTransport.publishMessages>
               <topic>{$ctx:topic}</topic>
               <key>{$ctx:key}</key>
               <value>{$ctx:value}</value>
               <valueSchema>{$ctx:valueSchema}</valueSchema>
            </kafkaTransport.publishMessages>
        </inSequence>
        <outSequence/>
        <faultSequence/>
    </resource>
</api>
```

> In the above example the <b>basicAuthCredentialsSource</b> parameter is configured to <b>USER_INFO</b>. If <b>basicAuthCredentialsSource</b> parameter is set to URL as,

Reviewer: ```suggestion
In the above example, the <b>basicAuthCredentialsSource</b> parameter is configured as <b>USER_INFO</b>. For example, consider a scenario where the <b>basicAuthCredentialsSource</b> parameter is set to <b>URL</b> as follows:
```

````xml 
<basicAuthCredentialsSource>URL</basicAuthCredentialsSource>
````

> then the <b>schemaRegistryUrl</b> parameter should be configured as shown below,

Reviewer: ```suggestion
Then, the <b>schemaRegistryUrl</b> parameter should be configured as shown below.
```

````xml 
<schemaRegistryUrl>http://admin:admin@localhost:8081</schemaRegistryUrl>
````
You may refer to the [confluent documentation](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/serdes-avro.html) for more details.

> This demonstrates how the Kafka connector publishes avro messages to the Kafka brokers.

Reviewer: ```suggestion
This demonstrates how the Kafka connector publishes Avro messages to Kafka brokers.
```
